miasm2/analysis/mem.py

```diff
@@ -0,0 +1,1259 @@
+    - Array: a fixed size array of MemFields (points)
+    - Union: similar to `union` in C, list of MemFields at the same offset in a
+      structure; the union has the size of the biggest MemField
+    - BitField: similar to C bitfields, a list of
+      [(<field_name), (number_of_bits)]; creates fields that correspond to
+      certain bits of the field
+
+A MemField always has a fixed size in memory.
+
+
+Some special memory structures are already implemented; they all are subclasses
+of MemStruct with a custom implementation:
+
+    - MemSelf: this class is just a special marker to reference a MemStruct
+      subclass inside itself. Works with Ptr and Array (e.g. Ptr(_, MemSelf)
+      for a pointer the same type as the class who uses this kind of field)
+    - MemVoid: empty MemStruct, placeholder to be casted to an implemented
+      MemStruct subclass
+    - MemStr: represents a string in memory; the encoding can be passed to the
+      constructor (null terminated ascii/ansi or null terminated utf16)
+    - MemArray: an unsized array of MemField; unsized here means that there is
+      no defined sized for this array, equivalent to a int* or char*-style table
+      in C. It cannot be allocated automatically, since it has no known size
+    - MemSizedArray: a sized MemArray, can be automatically allocated in memory
+      and allows more operations than MemArray
+    - mem: a function that dynamically generates a MemStruct subclass from a
+      MemField. This class has only one field named "value".
+
+A MemStruct do not always have a static size (cls.sizeof()) nor a dynamic size
+(self.get_size()).
+"""
+
+import logging
+import struct
+
+log = logging.getLogger(__name__)
+console_handler = logging.StreamHandler()
+console_handler.setFormatter(logging.Formatter("%(levelname)-5s: %(message)s"))
+log.addHandler(console_handler)
+log.setLevel(logging.WARN)
+
+# ALLOCATOR is a function(vm, size) -> allocated_address
+ALLOCATOR = None
+
+# Cache for dynamically generated MemStructs
+DYN_MEM_STRUCT_CACHE = {}
+
+def set_allocator(alloc_func):
+    """Set an allocator for this module; allows to instanciate statically sized
+    MemStructs (i.e. sizeof() is implemented) without specifying the address
+    (the object is allocated by @alloc_func in the vm.
+
+    Args:
+        alloc_func: func(VmMngr) -> integer_address
+    """
+    global ALLOCATOR
+    ALLOCATOR = alloc_func
+
+
+# Helpers
+
+def indent(s, size=4):
+    """Indent a string with @size spaces"""
+    return ' '*size + ('\n' + ' '*size).join(s.split('\n'))
+
+
+# FIXME: copied from miasm2.os_dep.common and fixed
+def get_str_ansi(vm, addr, max_char=None):
+    """Get a null terminated ANSI encoded string from a VmMngr.
+
+    Args:
+        vm: VmMngr instance
+        max_char: max number of characters to get in memory
+    """
+    l = 0
+    tmp = addr
+    while ((max_char is None or l < max_char) and
+           vm.get_mem(tmp, 1) != "\x00"):
+        tmp += 1
```

Comment:
If you plan to rewrite this function:

Actually, `tmp` is always equal to `addr + l`, so this variable is useless.
Additionally, there is a double memory access (one for *strlen*, the second one for getting the actual string), which may not reflect the reality, and could be inefficient.
You may rewrite it with a list joined at the end.